antlr4_11: init

Technius opened the pull request:
###### Description of changes

This adds ANTLR 4.11, which was released recently. The package was largely copied from `antlr4_9`, with a few modifications to account for the dependency changes from 4.9 to 4.11.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Technius committed:
antlr4_11: init

Technius commented on pkgs/development/tools/parsing/antlr/4.11.nix:
```diff
@@ -0,0 +1,96 @@
+	"-DANTLR4_INSTALL=on"
+
+        # Disable tests, since they require downloading googletest, which is
+	# not available in a sandboxed build.
```
Oops, noticed that there are some formatting issues here. Will fix tomorrow.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1209

SuperSandro2000 commented:
> This adds ANTLR 4.11, which was released recently. The package was largely copied from `antlr4_9`, with a few modifications to account for the dependency changes from 4.9 to 4.11.

What hinders us from upgrading the antlr version instead of adding a new one?

SuperSandro2000 reviewed (commented):
Can we please use some generic or common nix file to reduce the amount of code duplication?

Technius commented:
> What hinders us from upgrading the antlr version instead of adding a new one?

As far as I'm aware, ANTLR major/minor releases only provide best-effort backwards/forwards compatibility with no guarantees. For example, lexers/parsers generated by ANTLR 4.10 are incompatible with the ANTLR 4.9 runtime. Regarding the C++ runtime library specifically, there are also many other breaking changes between versions such as changes to the API, CMake configuration changes, etc.

Are you suggesting removing the versioned antlr packages and only leaving `pkgs.antlr4`, or changing `pkgs.antlr4` to point to ANTLR 4.11?

> Can we please use some generic or common nix file to reduce the amount of code duplication?

I agree there is a large amount of code duplication, but this PR is focused on adding ANTLR 4.11. Would it be better to refactor all of the ANTLR packages first, and include those refactoring changes + the new version in this PR, or would a separate refactoring PR be more appropriate?

SuperSandro2000 commented:
> or changing `pkgs.antlr4` to point to ANTLR 4.11?

we should think about that and removing older versions.

> Would it be better to refactor all of the ANTLR packages first, and include those refactoring changes + the new version in this PR,

That would be really nice to reduce the code duplication.

Technius committed:
antlr4_8, antlr4_9, antlr4_11: refactor definition

This commit refactors the antlr 4.8, 4.9, and 4.11 derivations to use common
code, greatly reducing the amount of code duplication, but at the expense of
slightly increased complexity.

aaronjheng commented:
It would be nice to have `4.10`.

Technius committed:
antlr4_10: init

SuperSandro2000 commented on pkgs/development/tools/parsing/antlr/4.nix:
```diff
@@ -0,0 +1,157 @@
+      sourceSha256 = "1af3cfqwk7lq1b5qsh1am0922fyhy7wmlpnrqdnvch3zzza9n1qm";
+      jarSha256 = "0dnz2x54kigc58bxnynjhmr5iq49f938vj6p50gdir1xdna41kdg";
+    }).extend (self: super: {
+      runtime.cpp = super.runtime.cpp.overrideAttrs (attrs: {
+        patchFlags = [ "-p3" ];
+        buildInputs = [ utf8cpp ]
+          ++ lib.optional stdenv.isLinux libuuid
+          ++ attrs.buildInputs;
+      });
+    })
```
Instead of doing overlays we should add an extra input the to generic packages where we can append buildInputs.

SuperSandro2000 commented on pkgs/development/tools/parsing/antlr/4.nix:
```diff
@@ -0,0 +1,157 @@
+        buildInputs =
+          lib.optional stdenv.isDarwin darwin.apple_sdk.frameworks.CoreFoundation;
+
+        postUnpack = ''
+          export sourceRoot=$sourceRoot/runtime/Cpp
+        '';
```
Can't we do ``sourceRoot = ......``?

SuperSandro2000 commented on pkgs/development/tools/parsing/antlr/4.nix:
```diff
@@ -0,0 +1,157 @@
+
+let
+
+  mkAntlr = { version, sourceSha256, jarSha256 }: lib.fixedPoints.makeExtensible (self: {
```
```suggestion
  mkAntlr = { version, sourceSha256, jarSha256 }: {
```
Do we really need this? Usually this is enough.

SuperSandro2000 commented on pkgs/development/tools/parsing/antlr/4.nix:
```diff
@@ -0,0 +1,157 @@
+        outputs = [ "out" "dev" "doc" ];
+
+        nativeBuildInputs = [ cmake ninja pkg-config ];
+        buildInputs =
+          lib.optional stdenv.isDarwin darwin.apple_sdk.frameworks.CoreFoundation;
```
```suggestion
        buildInputs = lib.optional stdenv.isDarwin CoreFoundation;
```
normally we inherit this in top-level.

Technius committed:
antlr4: simplify package definitions

This does not require a rebuild.

Technius committed:
antlr4: remove redundant settings

- Remove unneeded patch flags
- Move sourceRoot out of postUnpack

This change requires a rebuild.
